pkg/operator/configobservation/controllers/capability_imageregistry.go

```diff
@@ -0,0 +1,50 @@
+package controllers
+
+import (
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	openshiftcontrolplanev1 "github.com/openshift/api/openshiftcontrolplane/v1"
+	"github.com/openshift/cluster-openshift-controller-manager-operator/pkg/operator/configobservation"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+)
+
+func disabledImageRegistryControllers(listers configobservation.Listers) ([]openshiftcontrolplanev1.OpenShiftControllerName, error) {
+	cv, err := listers.ClusterVersionLister.Get("version")
+	if err != nil {
+		return nil, err
+	}
+	var imageRegistryCapabilityEnabled bool
+	for _, capability := range cv.Status.Capabilities.EnabledCapabilities {
+		if capability == configv1.ClusterVersionCapabilityImageRegistry {
+			imageRegistryCapabilityEnabled = true
+			break
+		}
+	}
+	controllers := []openshiftcontrolplanev1.OpenShiftControllerName{
+		openshiftcontrolplanev1.OpenShiftServiceAccountPullSecretsController,
+	}
+	if !imageRegistryCapabilityEnabled {
+		return controllers, nil
+	}
+
+	co, err := listers.ClusterOperatorLister.Get("image-registry")
```

Comment:
Why do we need those extra checks for internal image registry? We should not look into other operators information, only capabilities should drive the behavior.

Reply:
I'd also suggest squashing those 3 (or at least 2) identical methods into a single function, and a map holding information which capabilities match which controllers. This way you'll have only a single method passing a capability and in return you'll get a list of controllers to enable.

Reply:
> I'd also suggest squashing those 3 (or at least 2) identical methods into a single function, and a map holding information which capabilities match which controllers. This way you'll have only a single method passing a capability and in return you'll get a list of controllers to enable.

I used different methods for readability and maintainability as they represented three capabilities across three teams, and one of the cases was more complicated and needed more logic to compute.